lightning-invoice/src/payment.rs

```diff
@@ -466,6 +494,46 @@ where
 		}.map_err(|e| PaymentError::Sending(e))
 	}
 
+	fn process_path_inflight_htlcs(&self, hops: &Vec<RouteHop>) {
+		// total_inflight_map needs to be direction-sensitive when keeping track of the HTLC value
+		// that is held up. However, the `hops` array, which is a path returned by `find_route` in
+		// the router excludes the payer node. In the following lines, the payer's information is
+		// hardcoded with an inflight value of 0 so that we can correctly represent the first hop
+		// in our sliding window of two.
+		let our_node_id: PublicKey = self.payer.node_id();
+		let reversed_hops_with_payer = core::iter::once((0u64, our_node_id)).chain(
+			hops.split_last().unwrap().1.iter().map(|hop| (hop.fee_msat, hop.pubkey))).rev();
+		let mut cumulative_msat = hops.last().unwrap().fee_msat;
+
+		// Taking the reversed vector from above, we zip it with just the reversed hops list to
+		// work "backwards" of the given path, since the last hop's `fee_msat` actually represents
+		// the total amount sent.
+		for (_, (next_hop, prev_hop)) in hops.iter().rev().zip(reversed_hops_with_payer).enumerate() {
+			cumulative_msat += prev_hop.0;
+			self.total_inflight_map.lock().unwrap()
```

Comment:
We'll likely have to do something similar to scoring where we have a `LockableInflightHtlcs` trait and some `InflightHtlcs` struct (though no need for an associated type and trait). That way, when passing this to `find_route` we'll pass the locked version such that it sees a consistent view of the in-flight HTLCs.